File: app/src/main/java/fr/piotr/dismoitoutsms/util/ConfigurationManager.java
package fr.piotr.dismoitoutsms.util;

import android.content.Context;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;

import fr.piotr.dismoitoutsms.contacts.Contact;
import fr.piotr.dismoitoutsms.contacts.Contacts;

/**
 * @author Piotr
 * 
 */
public class ConfigurationManager {

	public enum Configuration {
		EMOTICONES, COMMANDE_VOCALE, UNIQUEMENT_CONTACTS, CONTACTS_BANNIS, LANGUE_DICTION, ARRET_STEP_DETECTOR, HEADSET_MODE
	}

	private static Map<String, String>	configuration	= new HashMap<>();

	private static final String			FICHIER			= "DisMoisToutSms.properties";

	public static void set(Context context, Configuration conf, String valeur) {
		charger(context);
		configuration.put(conf.toString(), valeur);
		sauvegarder(context);
	}

	public static void setBoolean(Context context, Configuration conf, boolean valeur) {
		set(context, conf, String.valueOf(valeur));
	}

	private static void sauvegarder(Context context) {
		FileOutputStream ous = null;
		OutputStreamWriter osw = null;
		try {
			// Crée un flux de sortie vers un fichier local.
			ous = context.openFileOutput(FICHIER, Context.MODE_PRIVATE);
			osw = new OutputStreamWriter(ous);
			for (final Entry<String, String> entry : configuration.entrySet()) {
				osw.write(entry.getKey() + "=" + entry.getValue() + "\n");
			}
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			try {
				if (osw != null) {
					osw.close();
				}
				if (ous != null) {
					ous.close();
				}
			} catch (IOException e) {
				e.printStackTrace();
			}
		}

	}

	private static void charger(Context context) {
		FileInputStream ins = null;
		InputStreamReader isr = null;
		BufferedReader br = null;
		try {
			// Crée un flux d’entrée depuis un fichier local.
			ins = context.openFileInput(FICHIER);
			isr = new InputStreamReader(ins);
			br = new BufferedReader(isr);
			String line = br.readLine();
			while (line != null) {
				String[] split = line.split("=");
				String cle = split[0];
				String valeur = "";
				if (split.length == 2) {
					valeur = split[1];
				}
				configuration.put(cle, valeur);
				line = br.readLine();
			}
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			try {
				if (br != null) {
					br.close();
				}
				if (isr != null) {
					isr.close();
				}
				if (ins != null) {
					ins.close();
				}
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	public static String get(Context context, Configuration conf) {
		charger(context);
		String valeur = configuration.get(conf.toString());
		if (valeur == null) {
			valeur = "";
		}
		return valeur;
	}

	public static boolean getBoolean(Context context, Configuration config) {
		return Boolean.TRUE.toString().equals(get(context, config));
	}

	public static boolean leContactEstBannis(Context context, Integer id) {
		String valeur = get(context, Configuration.CONTACTS_BANNIS);
		if (valeur == null) {
			return false;
		}
		List<String> bannis = Arrays.asList(valeur.split(";"));
		for (String cle : bannis) {
			if (cle.equals(id.toString())) {
				return true;
			}
		}
		return false;
	}

	public static void bannirContact(Context context, Integer id) {
		if (!leContactEstBannis(context, id)) {
			String valeur = get(context, Configuration.CONTACTS_BANNIS);
			valeur = valeur + id + ";";
			set(context, Configuration.CONTACTS_BANNIS, valeur);
		}
	}

	public static List<String> getIdsContactsBannis(Context context) {
		String valeur = get(context, Configuration.CONTACTS_BANNIS);
		if (valeur == null) {
			valeur = "";
		}
		return new ArrayList<>(Arrays.asList(valeur.split(";")));
	}

//	public static void gratifierContact(Context context, Integer id) {
//		if (leContactEstBannis(context, id)) {
//			String valeur = get(context, Configuration.CONTACTS_BANNIS);
//			List<String> bannis = new ArrayList<String>(Arrays.asList(valeur.split(";")));
//			bannis.remove(id.toString());
//			StringBuffer sb = new StringBuffer();
//			for (String cle : bannis) {
//				sb.append(cle);
//				sb.append(";");
//			}
//			valeur = sb.toString();
//			set(context, CONTACTS_BANNIS, valeur);
//		}
//	}
//
//	public static boolean aucunContactBanni(Context context) {
//		String valeur = get(context, Configuration.CONTACTS_BANNIS);
//		return valeur == null || valeur.isEmpty();
//	}
//
//	public static void gratifierTousLesContacts(Context context) {
//		set(context, CONTACTS_BANNIS, "");
//	}
//
//	public static void bannirTousLesContacts(Context context) {
//		Collection<Contact> allContacts = getAllContacts(context);
//		for (Contact contact : allContacts) {
//			bannirContact(context, contact.getId());
//		}
//	}

	public static void bannirLesContacts(Context context, Contacts contactsABannir) {
		StringBuilder valeur = new StringBuilder();
		for (Contact contact : contactsABannir) {
			valeur.append(contact.getId());
			valeur.append(";");
		}
		set(context, Configuration.CONTACTS_BANNIS, valeur.toString());
	}

	public static Locale getLangueSelectionnee(Context context) {
		String string = get(context, Configuration.LANGUE_DICTION);
		if (string == null || string.isEmpty()) {
			return Locale.getDefault();
		}
		Locale[] availableLocales = Locale.getAvailableLocales();
		for (Locale locale : availableLocales) {
			if (locale.getDisplayName().equals(string)) {
				return locale;
			}
		}
		return null;
	}

	public static void setLangueSelectionnee(Context context, String string) {
		Locale[] availableLocales = Locale.getAvailableLocales();
		for (Locale locale : availableLocales) {
			if (locale.getDisplayName().equals(string)) {
				set(context, Configuration.LANGUE_DICTION, string);
			}
		}
	}

}

File: app/src/main/java/fr/piotr/dismoitoutsms/util/ConfigurationManager.kt
package fr.piotr.dismoitoutsms.util

import android.content.Context
import android.text.TextUtils
import android.util.Log
import fr.piotr.dismoitoutsms.DisMoiToutSmsApplication
import fr.piotr.dismoitoutsms.contacts.Contacts
import java.io.*
import java.util.*

/**
 * @author Piotr
 */
object ConfigurationManager {

    private val configuration = HashMap<String, String>()
    private const val TAG = "ConfigurationManager"
    private const val FICHIER = "DisMoisToutSms.properties"
    private const val SEPARATOR = ";"

    val idsContactsBannis: List<String>
        get() {
            val context = DisMoiToutSmsApplication.INSTANCE.applicationContext
            var valeur: String? = getValue(context, Configuration.CONTACTS_BANNIS)
            if (valeur == null) {
                valeur = ""
            }
            return ArrayList(Arrays.asList(*valeur.split(SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()))
        }

    enum class Configuration {
        //@formatter:off
        TUTORIAL_DONE,
        EMOTICONES,
        COMMANDE_VOCALE,
        UNIQUEMENT_CONTACTS,
        CONTACTS_BANNIS,
        LANGUE_DICTION,
        HEADSET_MODE,
        BLUETOOTH_HEADSET_MODE,
        PRIVATE_LIFE_MODE,
        BLUETOOTH_DEVICES_BANNED
        //@formatter:on
    }

    private fun setValue(context: Context, conf: Configuration, valeur: String) {
        charger(context)
        configuration[conf.toString()] = valeur
        sauvegarder(context)
    }

    @JvmStatic
    fun setBoolean(context: Context, conf: Configuration, valeur: Boolean) {
        setValue(context, conf, valeur.toString())
    }

    private fun sauvegarder(context: Context) {

        var ous: FileOutputStream? = null
        var osw: OutputStreamWriter? = null
        try {
            // Crée un flux de sortie vers un fichier local.
            ous = context.openFileOutput(FICHIER, Context.MODE_PRIVATE)
            osw = OutputStreamWriter(ous!!)
            for ((key, value) in configuration) {
                osw.write("$key=$value\n")
            }
        } catch (e: IOException) {
            Log.e(TAG, e.message, e)
        } finally {
            osw?.close()
            ous?.close()
        }

    }

    private fun charger(context: Context) {
        var ins: FileInputStream? = null
        var isr: InputStreamReader? = null
        var br: BufferedReader? = null
        try {
            // Crée un flux d’entrée depuis un fichier local.
            ins = context.openFileInput(FICHIER)
            isr = InputStreamReader(ins!!)
            br = BufferedReader(isr)
            var line: String? = br.readLine()
            while (line != null) {
                val split = line.split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val cle = split[0]
                var valeur = ""
                if (split.size == 2) {
                    valeur = split[1]
                }
                configuration[cle] = valeur
                line = br.readLine()
            }
        } catch (e: IOException) {
            Log.e(TAG, e.message, e)
        } finally {
            try {
                br?.close()
                isr?.close()
                ins?.close()
            } catch (e: IOException) {
                Log.e(TAG, e.message, e)
            }

        }
    }

    private fun getValue(context: Context, conf: Configuration): String {
        charger(context)
        var valeur: String? = configuration[conf.toString()]
        if (valeur == null) {
            valeur = ""
        }
        return valeur
    }

    @JvmStatic
    fun getBoolean(context: Context, config: Configuration): Boolean {
        return java.lang.Boolean.TRUE.toString() == getValue(context, config)
    }

    fun bannirLesContacts(context: Context, contactsABannir: Contacts) {
        val valeur = StringBuilder()
        for ((id) in contactsABannir) {
            valeur.append(id)
            valeur.append(SEPARATOR)
        }
        setValue(context, Configuration.CONTACTS_BANNIS, valeur.toString())
    }

    @JvmStatic
    fun leContactEstBannis(context: Context, id: Int?): Boolean {
        val valeur = getValue(context, Configuration.CONTACTS_BANNIS)
        val bannis = Arrays.asList<String>(*valeur.split(";".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray())
        for (cle in bannis) {
            if (cle == id!!.toString()) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun getLangueSelectionnee(context: Context): Locale? {
        val string = getValue(context, Configuration.LANGUE_DICTION)
        if (string.isEmpty()) {
            return Locale.getDefault()
        }
        val availableLocales = Locale.getAvailableLocales()
        for (locale in availableLocales) {
            if (locale.displayName == string) {
                return locale
            }
        }
        return null
    }

    fun setLangueSelectionnee(context: Context, string: String) {
        val availableLocales = Locale.getAvailableLocales()
        for (locale in availableLocales) {
            if (locale.displayName == string) {
                setValue(context, Configuration.LANGUE_DICTION, string)
            }
        }
    }

    fun initBluetoothHeadsetModeFormigration(context: Context) {
        val value = getValue(context, Configuration.BLUETOOTH_HEADSET_MODE)
        if (TextUtils.isEmpty(value)) {
            setBoolean(context, Configuration.BLUETOOTH_HEADSET_MODE, getBoolean(context, Configuration.HEADSET_MODE))
        }
    }

    private fun getBluetoothBanned(context: Context): MutableList<String> {
        var valeur: String? = getValue(context, Configuration.BLUETOOTH_DEVICES_BANNED)
        if (valeur == null) {
            valeur = ""
        }
        return ArrayList(Arrays.asList(*valeur.split(SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()))
    }

    @JvmStatic
    fun isBluetoothBanned(context: Context, address: String): Boolean {
        return getBluetoothBanned(context).contains(address)
    }

    private fun banBluetoothDevice(context: Context, addressToBan: String) {
        val bluetoothBanned = getBluetoothBanned(context)
        if (!bluetoothBanned.contains(addressToBan)) {
            bluetoothBanned.add(addressToBan)
        }
        val valeur = StringBuilder()
        for (address in bluetoothBanned) {
            valeur.append(address)
            valeur.append(SEPARATOR)
        }
        setValue(context, Configuration.BLUETOOTH_DEVICES_BANNED, valeur.toString())
    }

    private fun grantBluetoothDevice(context: Context, addressToBan: String) {
        val bluetoothBanned = getBluetoothBanned(context)
        if (bluetoothBanned.contains(addressToBan)) {
            bluetoothBanned.remove(addressToBan)
        }
        val valeur = StringBuilder()
        for (address in bluetoothBanned) {
            valeur.append(address)
            valeur.append(SEPARATOR)
        }
        setValue(context, Configuration.BLUETOOTH_DEVICES_BANNED, valeur.toString())
    }

    @JvmStatic
    fun toggleBluetoothDevice(context: Context, address: String, checked: Boolean) {
        if (checked) {
            grantBluetoothDevice(context, address)
        } else {
            banBluetoothDevice(context, address)
        }
    }

}
